### PR TITLE
Add new `add_as_inner_object` option to add metadata to json/xml documents

### LINKED ISSUE #237: Fscrawler does not delete documents from index on json support
Tested on version 2.2. Setting json_support=true, the crawler does not manage to delete documents from the index.

### PR DESCRIPTION
**Pull request by babadofar**
This works for me, the tests are also running. 
Fixes #237 

This PR makes all content types have the same set of metadata, most importantly it adds info on file paths to json and xml documents allowing checking for deletions/updates. Perhaps there should be distinct indexing types, since there will be empty elements ( tika processed files will not have json_content, and json/xml files do not populate the meta object)

This is probably a breaking change, since it moves the source content of indexed json and xml documents into a sub element, aptly named "json_content". (sorry about that, xml)

### CODE REVIEW

**Comment by babadofar** [comment]
Actually trying to run the tests revealed that some of them failed. Tests should be fixed now, tested running from Intellij. However I get errors when running tests from maven. Not sure why, but I suppose that would be another issue.

**Comment by babadofar** [comment]
Any comments on this @dadoonet ?:)

**Comment by dadoonet** [comment]
@babadofar I'm very sorry. It's on my todo as soon as I'll have quiet time.

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java [hunk 1]
```diff
@@ -556,7 +554,25 @@ private void indexFile(FileAbstractModel fileAbstractModel, ScanStatistic stats,
             try {
                 // Create the Doc object
                 Doc doc = new Doc();
+                String docId = new String();
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java]:
```java
String docId;
```

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java [hunk 2]
```diff
@@ -556,7 +554,25 @@ private void indexFile(FileAbstractModel fileAbstractModel, ScanStatistic stats,
...
+                if(fsSettings.getFs().isFilenameAsId()){
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java]: Just formatting issue:

```java
if (fsSettings.getFs().isFilenameAsId()) {
```

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java [hunk 3]
```diff
@@ -582,33 +598,18 @@ private void indexFile(FileAbstractModel fileAbstractModel, ScanStatistic stats,
...
                 // We index
                 esIndex(fsSettings.getElasticsearch().getIndex(),
                         fsSettings.getElasticsearch().getType(),
-                        SignTool.sign((new File(filepath, filename)).toString()),
+                        docId,
                         doc);
-            } finally {
+
+            }
+            catch (Exception e){
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java]: Please remove the catch block here

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/meta/doc/Doc.java [hunk 4]
```diff
@@ -29,13 +31,17 @@
...
     private Attributes attributes;
+    private Map jsonContent;
 
     public Doc() {
         meta = new Meta();
         file = new File();
         path = new Path();
     }
 
+    public Map getJsonContent(){return jsonContent;}
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/meta/doc/Doc.java]: Formatting issue. Please write:

```java
    public Map getJsonContent() {
        return jsonContent;
    }

    public void setJsonContent(Map jsonContent ) {
        this.jsonContent = jsonContent;
    }
```

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/meta/doc/DocParser.java [hunk 5]
```diff
@@ -33,4 +34,7 @@ public static String toJson(Doc doc) throws JsonProcessingException {
     public static Doc fromJson(String json) throws IOException {
         return prettyMapper.readValue(json, Doc.class);
     }
+    public static Map fromJsonToMap(String json) throws IOException {
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/meta/doc/DocParser.java]: Can you add an empty line just before this please?

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClient.java [hunk 6]
```diff
@@ -52,6 +52,7 @@
 
     private final RestClient client;
     private String FIELDS = null;
+    private Integer defaultSearchSize = 10;
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClient.java]:
```java
public static final Integer defaultSearchSize = 10;
```

File: src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllDocumentsIT.java [hunk 7]
```diff
@@ -184,7 +188,7 @@ private SearchResponse runSearch(String filename, String content) throws IOExcep
         if (content != null) {
             fullQuery += " +content:" + content;
         }
-        SearchResponse response = elasticsearchClient.search("fscrawler_test_all_documents", null, fullQuery, null, "*");
+        SearchResponse response = elasticsearchClient.search("fscrawler_test_all_documents", null, fullQuery, 10, "*");
```
Comment (dadoonet) [inline src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllDocumentsIT.java]: Use here `ElasticsearchClient.defaultSearchSize` instead?

**Comment by dadoonet** [comment]
@babadofar Ha sorry. I pushed new commits which creates again some conflicts with your branch. :(

**Comment by babadofar** [comment]
I merged in latest changes from master, and, all tests are green on my machine now :)

I have not yet implemented a way to select whether you want support for deletes and updates of json and xml documents or not.  Should it be an option in the config file do you think?

**Comment by dadoonet** [comment]
Yes. There is a ignore delete option which should be set to true if this new option is defined/undefined.

**Comment by babadofar** [comment]
Ok, I'll have to look into it another day. Should it be named  something other than remove_deleted? Since disabling this would also mean to ignore updates?

**Comment by dadoonet** [comment]
> Should it be named something other than remove_deleted? Since disabling this would also mean to ignore updates?

Well. Why not but in that case please keep the old method around, mark it as Deprecated and add a method like `FsCrawler#moveLegacyResources` which reads an old format and rewrite the job file using the new setting instead.

So basically:

* move `Fs.removeDeleted` as a `Boolean` instead of `boolean`
* if `getRemoveDeleted() != null`, then `setNewName(getRemoveDeleted())` and `setRemoveDeleted(null)`
* save the job file

But I'm afraid this option `ignore_updates` would become confusing.
Because indeed it would not update JSON docs but binary docs would be still updated. Or we have to change the logic and check in ES if a document already exists then ignore it if `ignore_updates` is true. Which is IMO a way too much complicated.

So thinking about it, I'd keep for now `remove_deleted` as it is but may be add a log.info if the new option is not set and remove deleted is true. 

WDYT?

**Comment by babadofar** [comment]
Yes, I see the difficulties here if you want to maintain backward compatibility. There is also the problem with the other config settings such as metadata and attributes on JSON docs which are not saved to index. 
I merged in the lastest changes from master, and added an option that should allow using the deprecated settings, However, I did not have the time yet to check properly that it works. Sorry about this taking so long, but there's only so little time, if you know what i mean:)

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClient.java [hunk 8]
```diff
@@ -55,6 +55,7 @@
     private final Elasticsearch settings;
     private final RestClient client;
     private String FIELDS = null;
+    public static final Integer defaultSearchSize = 10;
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClient.java]: Can we write it as `DEFAULT_SEARCH_SIZE`?

Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClient.java]: That being said, it's only used in tests once so may be just change the test and use `10`?
That said your change:

```java
SearchResponse response = elasticsearchClient.search("fscrawler_test_all_documents", null, fullQuery, ElasticsearchClient.defaultSearchSize, null);
```

Is not really related to this PR, right?

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClientManager.java [hunk 9]
```diff
@@ -95,12 +95,14 @@ public void createIndexAndMappings(FsSettings settings, boolean updateMapping) t
         try {
             // If needed, we create the new mapping for files
             Path jobMappingDir = config.resolve(settings.getName()).resolve("_mappings");
-            if (!settings.getFs().isJsonSupport() && !settings.getFs().isXmlSupport()) {
+       // if deprecated settings
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClientManager.java]: Indent?

File: .gitignore [hunk 10]
```diff
@@ -5,3 +5,5 @@
 .idea
 *.iml
 plugin_tools
+
+*.orig
```
Comment (babadofar) [inline .gitignore]: It's from git mergetool. It defaults to creating backup files after resolving a git conflict.

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClient.java (hunk 8 above)
Comment (babadofar) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClient.java]: I removed it

**Comment by babadofar** [comment]
I finally managed to incorporate the deprecated behaviour, and made it a configurable parameter. There is a test that demonstrates the functionality and I even tried to scribble down some documentation.

File: README.md [hunk 11]
```diff
@@ -431,6 +431,7 @@ Here is a list of Local FS settings (under `fs.` prefix)`:
 | `fs.excludes`                    | `null`        | [Includes and Excludes](#includes-and-excludes)                                   |
 | `fs.json_support`                | `false`       | [Indexing JSon docs](#indexing-json-docs)                                         |
 | `fs.xml_support`                 | `false`       | [Indexing XML docs](#indexing-xml-docs) (from 2.2)                                |
+| `fs.use_deprecated_json_setup`   | `false`       | [Use Deprecated JSon setup](#use-deprecated-json-setup)                                 |
```
Comment (dadoonet) [inline README.md]: Thanks for writing this doc. I'm very sorry to say that at the end of the day, I'd prefer a setting named `add_as_inner_object` which is `false` by default (so default existing behavior).

If `true`, then we send the inner content to `object` field (instead of `json_content`).

File: README.md [hunk 12]
```diff
@@ -517,7 +518,8 @@ Note that `includes` does not apply to directory names but only to filenames.
 #### Indexing JSon docs
 
 If you want to index JSon files directly without parsing with Tika, you can set `json_support` to `true`.
-
+The content of JSon files will be added below the field `json_content`.
```
Comment (dadoonet) [inline README.md]: s/json_content /object

File: README.md [hunk 13]
```diff
@@ -517,7 +518,8 @@ Note that `includes` does not apply to directory names but only to filenames.
...
+Set option [Use Deprecated JSon setup](#use-deprecated-json-setup) to store JSon contents directly under the _source element.
```
Comment (dadoonet) [inline README.md]: Change with `add_as_inner_object`.

File: README.md [hunk 14]
```diff
@@ -532,7 +534,8 @@ Of course, if you did not define a mapping before launching the crawler, Elastic
 #### Indexing XML docs
 
 If you want to index XML files and convert them to JSON, you can set `xml_support` to `true`.
-
+The content of XML files will be added below the field `json_content`. 
+Set option [Use Deprecated JSon setup](#use-deprecated-json-setup) to store XML contents directly under the _source element.
```
Comment (dadoonet) [inline README.md]: Change with `add_as_inner_object`.

File: README.md [hunk 15]
```diff
@@ -544,6 +547,20 @@ If you want to index XML files and convert them to JSON, you can set `xml_suppor
 
 Of course, if you did not define a mapping before launching the crawler, Elasticsearch will auto guess the mapping.
 
+#### Use Deprecated JSon Setup
```
Comment (dadoonet) [inline README.md]: Same goes here: we document instead the feature `add_as_inner_object`

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/meta/doc/Doc.java [hunk 16]
```diff
@@ -42,13 +44,22 @@
     private File file;
     private Path path;
     private Attributes attributes;
+    private Map jsonContent;
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/meta/doc/Doc.java]: object

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/meta/settings/Fs.java [hunk 17]
```diff
@@ -32,6 +32,7 @@
     private boolean filenameAsId;
     private boolean addFilesize = true;
     private boolean removeDeleted = true;
+    private boolean useDeprecatedJsonSetup = false;
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/meta/settings/Fs.java]: addAsInnerObject

File: src/main/resources/fr/pilato/elasticsearch/crawler/fs/_default/5/doc.json [hunk 18]
```diff
@@ -51,6 +51,9 @@
         }
       }
     },
+    "json_content" : {
```
Comment (dadoonet) [inline src/main/resources/fr/pilato/elasticsearch/crawler/fs/_default/5/doc.json]: object

File: src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/AbstractITCase.java [hunk 19]
```diff
@@ -244,6 +244,8 @@ public static SearchResponse countTestHelper(final String indexName, String quer
                 sr.setFields(fields);
             }
 
+            sr.setSize(0);
```
Comment (dadoonet) [inline src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/AbstractITCase.java]: I'd not set that. I think it can have bad impact on some tests which read the response sent by elasticsearch later on.

File: src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java [hunk 20]
```diff
@@ -466,7 +466,7 @@ public void test_json_support() throws Exception {
 
         assertThat("We should have 0 doc for tweet in text field...", awaitBusy(() -> {
             try {
-                SearchResponse response = elasticsearchClient.search(getCrawlerName(), null, "text:tweet");
+                SearchResponse response = elasticsearchClient.search(getCrawlerName(), null, "json_content.text:tweet");
```
Comment (dadoonet) [inline src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java]: object

File: src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java [hunk 21]
```diff
@@ -534,6 +534,24 @@ public void test_filename_as_id() throws Exception {
         }), equalTo(true));
     }
 
+    public void test_json_deprecated() throws Exception {
```
Comment (dadoonet) [inline src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java]: test_add_as_inner_object

File: src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java [hunk 22]
```diff
@@ -534,6 +534,24 @@ public void test_filename_as_id() throws Exception {
...
+        Fs fs = startCrawlerDefinition()
+                .setUseDeprecatedJsonSetup(true)
+                .build();
+        startCrawler(getCrawlerName(), fs, endCrawlerDefinition(getCrawlerName()), null);
+
+        assertThat("We should have 0 doc for tweet in text field...", awaitBusy(() -> {
```
Comment (dadoonet) [inline src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java]: Message is wrong as we are expecting 2 docs.

File: README.md (hunk 11 above)
Comment (babadofar) [inline README.md]: No problem :)

**Comment by babadofar** [comment]
The changes you requested should be implemented now. I also tried to rewrite the documentation to reflect the behaviour.

File: README.md [hunk 23]
```diff
@@ -550,6 +552,20 @@ If you want to index XML files and convert them to JSON, you can set `xml_suppor
 
 Of course, if you did not define a mapping before launching the crawler, Elasticsearch will auto guess the mapping.
 
+#### Add as Inner Object
```
Comment (dadoonet) [inline README.md]: Unsure if it's needed but can you add an empty line here?

File: README.md [hunk 24]
```diff
@@ -523,7 +524,7 @@ By default, FS crawler will exclude files starting with `~`.
 #### Indexing JSon docs
 
 If you want to index JSon files directly without parsing with Tika, you can set `json_support` to `true`.
-
+JSon contents will be stored directly under _source. If you need to keep JSon documents synchronized to the index, set option [Add as Inner Object](#add-as-inner-object) which stores additional metadata and the  JSon contents under field `object`.
```
Comment (dadoonet) [inline README.md]: Really small change: can you change `the  JSon` to `the JSon` (one single space).

File: README.md [hunk 25]
```diff
@@ -550,6 +552,20 @@ If you want to index XML files and convert them to JSON, you can set `xml_suppor
 
 Of course, if you did not define a mapping before launching the crawler, Elasticsearch will auto guess the mapping.
 
+#### Add as Inner Object
+The default settings store the contents of json and xml documents directly onto the _source element of elasticsearch documents. Thereby, there is no metadata about file and path settings, which are necessary to determine if a document is deleted or updated.
+New files will however be added to the index, (determined by the file timestamp).
+
+If you need to keep json or xml documents synchronized to elasticsearch, you should set this option.
+```json
```
Comment (dadoonet) [inline README.md]: Line break before?

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java [hunk 26]
```diff
@@ -544,7 +546,12 @@ private void indexFile(FileAbstractModel fileAbstractModel, ScanStatistic stats,
             try {
                 // Create the Doc object
                 Doc doc = new Doc();
-
+                String docId;
+                if (fsSettings.getFs().isFilenameAsId()) {
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java]: `generateIdFromFilename` does already this test I believe. This change is not needed anymore I think.

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java [hunk 27]
```diff
@@ -568,33 +575,40 @@ private void indexFile(FileAbstractModel fileAbstractModel, ScanStatistic stats,
...
-                        // Extracting content with Tika
                         generate(fsSettings, inputStream, filename, doc, messageDigest, filesize);
                     }
                 }
-
-                // We index
                 esIndex(esClientManager.bulkProcessor(), fsSettings.getElasticsearch().getIndex(),
                         fsSettings.getElasticsearch().getType(),
-                        generateIdFromFilename(filename, dirname),
+                        docId,
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java]: `generateIdFromFilename` does already what is needed. This change is not needed anymore I think.

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java [hunk 28]
```diff
@@ -568,33 +575,40 @@ private void indexFile(FileAbstractModel fileAbstractModel, ScanStatistic stats,
...
+                                    fsSettings.getElasticsearch().getType(),
+                                    generateIdFromFilename(filename, dirname),
+                                    XmlDocParser.generate(inputStream));
+                            return;
+                        } else {
+                                // https://github.com/dadoonet/fscrawler/issues/185 : Support Xml files
+                                doc.setObject(XmlDocParser.generateMap(inputStream));
+                        }
                     } else {
-                        // Extracting content with Tika
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java]: Why did you remove the comment?

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClientManager.java [hunk 29]
```diff
@@ -100,12 +100,13 @@ public void createIndexAndMappings(FsSettings settings, boolean updateMapping) t
 
         try {
             // If needed, we create the new mapping for files
-            if (!settings.getFs().isJsonSupport() && !settings.getFs().isXmlSupport()) {
+            if (settings.getFs().isAddAsInnerObject() == false ||  (!settings.getFs().isJsonSupport() && !settings.getFs().isXmlSupport())) {
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClientManager.java]: haha! That's definitely how elasticsearch policy wants negative tests to be implemented :)
You can keep `== false` or use `!` as you prefer.

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/meta/settings/FsSettings.java [hunk 30]
```diff
@@ -28,6 +28,7 @@
     private Elasticsearch elasticsearch;
     private Rest rest;
 
+
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/meta/settings/FsSettings.java]: Not needed.

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/XmlDocParser.java [hunk 31]
```diff
@@ -63,4 +63,16 @@ public static String generate(InputStream inputStream) throws IOException {
         logger.trace("Generated JSON: {}", json);
         return json;
     }
+
+
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/XmlDocParser.java]: Can you remove one of the empty lines?

File: src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllDocumentsIT.java [hunk 32]
```diff
@@ -113,6 +114,10 @@ public void testXmlIssue163() throws IOException {
     }
 
     @Test
+    public void testJson() throws IOException {
+        runSearch("test.json", "json");
+    }
+    @Test
```
Comment (dadoonet) [inline src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllDocumentsIT.java]: Can you add an empty line please?

File: src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java [hunk 33]
```diff
@@ -464,9 +464,9 @@ public void test_json_support() throws Exception {
                 .build();
         startCrawler(getCrawlerName(), fs, endCrawlerDefinition(getCrawlerName()), null);
 
-        assertThat("We should have 0 doc for tweet in text field...", awaitBusy(() -> {
+        assertThat("We should have 2 doc for tweet in text field...", awaitBusy(() -> {
```
Comment (dadoonet) [inline src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java]: s/text/object.text ?

File: src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java [hunk 34]
```diff
@@ -811,7 +829,7 @@ public void test_json_support_and_other_files() throws Exception {
         assertThat("We should have 2 docs only...", awaitBusy(() -> {
```
Comment (dadoonet) [inline src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java]: 2 or 3?

File: src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java [hunk 35]
```diff
@@ -525,6 +525,24 @@ public void test_filename_as_id() throws Exception {
         }), equalTo(true));
     }
 
+    public void test_add_as_inner_object() throws Exception {
```
Comment (dadoonet) [inline src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java]: I think there is something wrong with the implementation.
You are setting here `add_as_inner_object` to `true` but at the end the json document is not copied to the `object` field.

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java (hunk 26 above)
Comment (babadofar) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java]: ok

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java (hunk 28 above)
Comment (babadofar) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java]: no idea why?

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java (hunk 27 above)
Comment (babadofar) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java]: ok

File: src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java (hunk 35 above)
Comment (babadofar) [inline src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java]: haha, yes this should be the other way around :)

Comment (babadofar) [inline src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java]: No idea how I could believe that test worked. Should remember to stay off LSD while coding.

**Comment by babadofar** [comment]
I fixed the implementation, so it actually does what it says, added a test that even runs, and other stuff. Anything else?

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClientManager.java (hunk 29 above)
Comment (babadofar) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClientManager.java]: DIVERSITY!

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java [hunk 36]
```diff
@@ -568,30 +569,37 @@ private void indexFile(FileAbstractModel fileAbstractModel, ScanStatistic stats,
...
+                            // https://github.com/dadoonet/fscrawler/issues/5 : Support JSon files
+                            doc.setObject(DocParser.fromJsonToMap(read(inputStream)));
+                        } else {
+                            esIndex(esClientManager.bulkProcessor(), fsSettings.getElasticsearch().getIndex(),
+                                    fsSettings.getElasticsearch().getType(),
+                                    generateIdFromFilename(filename, dirname),
+                                    read(inputStream));
+                            return;
+                        }
+                    } else if (fsSettings.getFs().isXmlSupport() && fileExtension.equals("xml")) {
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/FsCrawlerImpl.java]: Same here. Don't test extension. If people want to mix different type of files, they can use `includes` parameter and launch multiple instances of FSCrawler.

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClientManager.java (hunk 29 above)
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/client/ElasticsearchClientManager.java]: Can you remove the extra space after `||`?

File: src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/XmlDocParser.java [hunk 37]
```diff
@@ -63,4 +63,15 @@ public static String generate(InputStream inputStream) throws IOException {
         logger.trace("Generated JSON: {}", json);
         return json;
     }
+
+    public static Map generateMap(InputStream inputStream) throws IOException {
+        logger.trace("Converting XML document [{}]");
+        // Extracting XML content
```
Comment (dadoonet) [inline src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/XmlDocParser.java]: May be move this comment as a Javadoc?

File: src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java [hunk 38]
```diff
@@ -833,6 +855,7 @@ public void test_xml_enabled() throws Exception {
         countTestHelper(getCrawlerName(), "title:maeve", 1);
         countTestHelper(getCrawlerName(), "price:[5 TO 6]", 2);
 
+
```
Comment (dadoonet) [inline src/test/java/fr/pilato/elasticsearch/crawler/fs/test/integration/FsCrawlerImplAllParametersIT.java]: Non needed empty line.

**Comment by dadoonet** [comment]
@babadofar Thanks a lot. I rewrote a bit your branch and the code and wrote a single commit instead 1acbd76
It's now pushed on master. Let's hope everything is right. :)
